test/smtcallback.js

```diff
@@ -138,82 +255,27 @@ tape('SMTCheckerCallback', function (t) {
-    var dirs = [testdir];
-    var i;
-    while (dirs.length > 0) {
-      var dir = dirs.shift();
-      var files = fs.readdirSync(dir);
-      for (i in files) {
-        var file = path.join(dir, files[i]);
-        if (fs.statSync(file).isDirectory()) {
-          dirs.push(file);
-        } else {
-          sources.push(file);
-        }
-      }
-    }
-
-    // Read tests and collect expectations
-    var tests = [];
-    for (i in sources) {
-      st.comment('Collecting ' + sources[i] + '...');
-      var source = fs.readFileSync(sources[i], 'utf8');
-
-      var engine;
-      var option = '// SMTEngine: ';
-      if (source.includes(option)) {
-        let idx = source.indexOf(option);
-        if (source.indexOf(option, idx + 1) !== -1) {
-          st.skip('SMTEngine option given multiple times.');
-          st.end();
-          return;
-        }
-        let re = new RegExp(option + '(\\w+)');
-        let m = source.match(re);
-        assert(m !== undefined);
-        assert(m.length >= 2);
-        engine = m[1];
-      }
-
-      var expected = [];
-      var delimiter = '// ----';
-      if (source.includes(delimiter)) {
-        expected = source.substring(source.indexOf('// ----') + 8, source.length).split('\n');
-        // Sometimes the last expectation line ends with a '\n'
-        if (expected.length > 0 && expected[expected.length - 1] === '') {
-          expected.pop();
-        }
-      }
-      tests[sources[i]] = {
-        expectations: expected,
-        solidity: { test: { content: preamble + source } },
-        ignoreCex: source.includes('// SMTIgnoreCex: yes'),
-        engine: engine
-      };
-    }
+    const sources = collectFiles(testdir);
+    const tests = createTests(sources, st);
 
     // Run all tests
-    for (i in tests) {
-      var test = tests[i];
-
-      // Z3's nondeterminism sometimes causes a test to timeout in one context but not in the other,
-      // so if we see timeout we skip a potentially misleading run.
-      var findError = (errorMsg) => { return errorMsg.includes('Error trying to invoke SMT solver'); };
-      if (test.expectations.find(findError) !== undefined) {
-        st.skip('Test contains timeout which may have been caused by nondeterminism.');
-        continue;
-      }
+    for (let i in tests) {
+      const test = tests[i];
+      st.comment('Running ' + i);
 
-      var settings = {};
+      let settings = {};
       // `pragma experimental SMTChecker;` was deprecated in 0.8.4
       if (semver.gt(solc.semver(), '0.8.3')) {
-        let engine = test.engine !== undefined ? test.engine : 'all';
+        const engine = test.engine !== undefined ? test.engine : 'all';
         settings = { modelChecker: { engine: engine } };
+
+        // 0.8.5 introduced the `solvers` option.
+        // We force the compiler to use smtlib2 in this test.
+        if (semver.gt(solc.semver(), '0.8.4')) {
+          settings.modelChecker.solvers = ['smtlib2'];
```

Comment:
Because of this, from 0.8.5 on we *don't* use the static z3 inside soljson.js for this test anymore, but require a z3 binary in the system.